kong/db/schema/entities/keys.lua

```diff
@@ -0,0 +1,128 @@
+-- This software is copyright Kong Inc. and its licensors.
+-- Use of the software is subject to the agreement between your organization
+-- and Kong Inc. If there is no such agreement, use is governed by and
+-- subject to the terms of the Kong Master Software License Agreement found
+-- at https://konghq.com/enterprisesoftwarelicense/.
+-- [ END OF LICENSE 0867164ffc95e54f04670b5169c09574bdbd9bba ]
+
+local typedefs = require "kong.db.schema.typedefs"
+local Schema = require "kong.db.schema"
+local cjson = require "cjson.safe"
+local supported_key_formats = require "kong.constants".KEY_FORMATS
+
+return {
+  name          = "keys",
+  dao           = "kong.db.dao.keys",
+  primary_key   = { "id" },
+  cache_key     = { "kid", "set" },
+  endpoint_key  = "name",
+  workspaceable = true,
+  ttl           = false,
+  fields        = {
+    {
+      id = typedefs.uuid,
+    },
+    {
+      set = {
+        type      = "foreign",
+        required  = false,
+        reference = "key_sets",
+        on_delete = "cascade",
+      },
+    },
+    {
+      name = {
+        type     = "string",
+        required = false,
+        unique   = true,
+      },
+    },
+    {
+      kid = {
+        type     = "string",
+        required = true,
+        unique   = true,
+      },
+    },
+    {
+      jwk = {
+        -- type string but validate against typedefs.jwk
+        type = "string",
+        referenceable = true,
+        encrypted = true
+      }
+    },
+    {
+      pem = typedefs.pem
+    },
+    {
+      tags = typedefs.tags,
+    },
+    {
+      created_at = typedefs.auto_timestamp_s,
+    },
+    {
+      updated_at = typedefs.auto_timestamp_s,
+    },
+  },
+  entity_checks = {
+    -- XXX: add mutually exclusive to jwk and pem for now.
+    -- to properly implement this we need to check that the keys are the same
+    -- to avoid false assumptions for an object.
+    {
+      mutually_exclusive = supported_key_formats
+    },
+    {
+      at_least_one_of = supported_key_formats
+    },
+    { custom_entity_check = {
+      field_sources = { "jwk", "pem", "kid" },
+      fn = function(entity)
+        -- JWK validation
+        if type(entity.jwk) == "string" then
+          if kong.vault.is_reference(entity.jwk) then
+            -- can't validate a reference
+            return true
+          end
+          -- validate against the typedef.jwk
+          local schema = Schema.new(typedefs.jwk)
+          if not schema then
+            return nil, "could not load jwk schema"
+          end
```

Comment:
typedefs.jwk should already be an instance of Schema, so invoking Schema.new again shouldn't be needed here (untested):

```suggestion
          local schema = typedefs.jwk
```

Reply:
`typedefs.jwk` is a schema _definition_ but not a Schema instance.